Implement z-scoring and gap optimization in CE Align

Will-Tyler opened the pull request:
### Acknowledgements

<!--- Please read each of the following items and confirm by replacing
 !--the [ ] with a [X] --->

- [X] I hereby agree to dual licence this and any previous contributions under both
the _Biopython License Agreement_ **AND** the _BSD 3-Clause License_.

- [X] I have read the ``CONTRIBUTING.rst`` file, have run ``pre-commit``
locally, and understand that continuous integration checks will be used to
confirm the Biopython unit tests and style checks pass with these changes.

- [X] I have added my name to the alphabetical contributors listings in the files
``NEWS.rst`` and ``CONTRIB.rst`` as part of this pull request, am listed
already, or do not wish to be listed. (*This acknowledgement is optional.*)

<!--- Briefly describe the changes included in this pull request below
 !--- starting with 'Closes #...' if appropriate --->

### Description

In this pull request, I implement z-score calculation and gap optimization in the PDB module's implementation of CE Align. The **z-score** is a measure of statistical significance or how likely a particular alignment would be encountered by chance. **Gap optimization** is a procedure that iteratively adjusts gaps in an alignment to improve the alignment's RMSD. In the CE Align algorithm, gap optimization is performed when the z-score is above a certain threshold (normally 3.5).

These changes improve the alignment by searching through alternative alignments when the alignment is statistically significant.

### Testing

I adjusted the unit tests for the improved RMSD, and I performed some manual testing.

I used the [MALIDUP](http://prodata.swmed.edu/malidup/) and [MALISAM](http://prodata.swmed.edu/malisam/) databases to manually test these changes.

Here is my testing script:
```python
import os
import time
from itertools import islice
import re
from Bio.PDB import PDBParser, CEAligner
from collections import deque
import pandas as pd


pattern = re.compile("^[a-z0-9]{4}_\\w\\d.pdb$")
parser = PDBParser()
aligner = CEAligner()
results = deque()

for dirpath, _, filenames in islice(os.walk("data/dup"), 1, None):
    dirname = os.path.split(dirpath)[-1]
    pdb_code = dirname[1:5]
    pdb_filenames = list(filename for filename in filenames if pattern.match(filename))

    assert len(pdb_filenames) == 2

    structure_1 = parser.get_structure(pdb_code, os.path.join(dirpath, pdb_filenames[0]))
    structure_2 = parser.get_structure(pdb_code, os.path.join(dirpath, pdb_filenames[1]))

    start_time = time.time()
    aligner.set_reference(structure_1)
    aligner.align(structure_2, transform=False)
    results.append({
        "PDB Code": pdb_code,
        "RMSD": aligner.rms,
        "Length": aligner.length,
        "Time": time.time() - start_time,
        "Did optimize?": aligner.did_optimize,
        "Z Score": aligner.z_score,
    })

results = pd.DataFrame(results)
print(results[["RMSD", "Length", "Time", "Did optimize?", "Z Score"]].describe())
print(results["Did optimize?"].describe())

pattern = re.compile("^[a-z0-9]{4}_\\w{2}.pdb$")
results = deque()

for dirpath, _, filenames in islice(os.walk("data/analogs"), 1, None):
    dirname = os.path.split(dirpath)[-1]
    pdb_filenames = list(filename for filename in filenames if pattern.match(filename))

    assert len(pdb_filenames) == 2

    structure_1 = parser.get_structure(pdb_filenames[0][:4], os.path.join(dirpath, pdb_filenames[0]))
    structure_2 = parser.get_structure(pdb_filenames[1][:4], os.path.join(dirpath, pdb_filenames[1]))

    start_time = time.time()
    aligner.set_reference(structure_1)
    aligner.align(structure_2, transform=False)
    results.append({
        "Directory": dirname,
        "RMSD": aligner.rms,
        "Length": aligner.length,
        "Time": time.time() - start_time,
        "Did optimize?": aligner.did_optimize,
        "Z Score": aligner.z_score,
    })

results = pd.DataFrame(results)
print(results[["RMSD", "Length", "Time", "Did optimize?", "Z Score"]].describe())
print(results["Did optimize?"].describe())
```

### Results with New Code

#### MALIDUP

```
             RMSD      Length        Time     Z Score
count  241.000000  241.000000  241.000000  241.000000
mean     3.287257   83.352697    0.024016    6.248091
std      1.269601   41.113411    0.025839    0.665213
min      0.628612   24.000000    0.001868    4.580000
25%      2.361698   56.000000    0.007987    5.860000
50%      3.189827   72.000000    0.016907    6.230000
75%      4.294839  104.000000    0.027342    6.700000
max      7.228586  256.000000    0.178572    7.930000
count      241
unique       1
top       True
freq       241
Name: Did optimize?, dtype: object
```

#### MALISAM

```
             RMSD      Length        Time    Z Score
count  130.000000  130.000000  130.000000  130.00000
mean     3.764365   61.292308    0.011495    6.69200
std      0.718092   12.932410    0.004768    0.40069
min      2.093043   24.000000    0.003485    5.60000
25%      3.212446   56.000000    0.008451    6.47000
50%      3.726223   56.000000    0.010203    6.70000
75%      4.314141   72.000000    0.013439    6.89250
max      5.365077  112.000000    0.033091    7.64000
count      130
unique       1
top       True
freq       130
Name: Did optimize?, dtype: object
```

### Results with Original Code

#### MALIDUP

```
             RMSD      Length        Time
count  241.000000  241.000000  241.000000
mean     3.449998   83.352697    0.008540
std      1.290120   41.113411    0.011653
min      0.628612   24.000000    0.000344
25%      2.570221   56.000000    0.001848
50%      3.416172   72.000000    0.004262
75%      4.406737  104.000000    0.010367
max      7.300840  256.000000    0.081144
```

#### MALISAM

```
             RMSD      Length        Time
count  130.000000  130.000000  130.000000
mean     3.946192   61.292308    0.004038
std      0.699545   12.932410    0.002931
min      2.480055   24.000000    0.000751
25%      3.409361   56.000000    0.002134
50%      3.872238   56.000000    0.003014
75%      4.476527   72.000000    0.005288
max      5.590935  112.000000    0.015195
```

### Discussion

The calculated z-score is above the threshold (3.5) and gap optimization is performed for all of the alignments in testing. This is unsurprising because the test alignments come from curated databases of homologous domains (MALIDUP) or structurally analogous motifs (MALISAM).

The z-score calculation and the gap optimization increase the total alignment time. However, I did not notice the difference in execution time without timing the test program. In homology searches, gap optimization should not be performed on every alignment because a z-score threshold of 3.5 corresponds to around 1 in 1000 alignments.

As expected, the alignment length is not affected by gap optimization. I included the alignment length as a simple sanity check.

Finally, CE Align with gap optimization does show better RMSDs.

The z-score calculation code is adapted from the [original CE Align code](https://github.com/kad-ecoli/CE). The gap optimization code is written by myself based on the description in the paper. I found it challenging to understand the gap optimization implementation in the original CE Align code.

I implement the z-score calculation in the C extension module because an alignment path's similarity metric is used to calculate the alignment path's z-score. The similarity metric is readily available in the extension module, and the original code was already available in C.

I implement gap optimization in Python because gap optimization is dependent on RMSD calculation, which is readily available in Python and not available in C.

### References

- [CE Align paper](https://pubmed.ncbi.nlm.nih.gov/9796821/)
- [Original CE Align code](https://github.com/kad-ecoli/CE)

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -211,6 +211,149 @@ calcS(
     return S;
 }
 
+static const double tableZtoP[] = {
```
Could I trouble you to leave a comment here explaining what P and Z are? I assume Z is just a Z-score, but what's P, a probability?

Will-Tyler replied:
Yes, that's correct. I'll add the comments.

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -211,6 +211,149 @@ calcS(
+    const int pathLength,
+    const double similarity)
+{
+    if (fragmentSize != 8 || pathLength < 1) {
```
This was only parameterized for the default window length right? I'd add a comment here to specify that.

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -211,6 +211,149 @@ calcS(
+    const int pathLength,
+    const int gapCount)
+{
+    if (fragmentSize != 8 || pathLength < 2) {
```
Why `pathLength < 2` here but above was `< 1`?

Will-Tyler replied:
That is how the [original source code](https://github.com/kad-ecoli/CE/blob/master/ce/ce_align.C#L1053) is. I did notice that the first element in the `gapCountAvgs` array is `0.00`. In the case where `pathLength` is `1`, the function still returns zero if the condition is changed to `pathLength < 1`. So i will go ahead and change it for consistency.

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -211,6 +211,149 @@ calcS(
+
+    double gapCountAvg, gapCountSD;
+
+    if (pathLength < 21) {
```
You might want to transfer these `21` to a static variable with a name to make it easier to understand. Something like `MIN_PATH_LENGTH`.

Will-Tyler replied:
I added a comment explaining that the number in the condition is the length of the arrays that store the averages and standard deviations. Let me know if this is good. I could also introduce a variable called `STATISTICS_ARRAYS_LENGTH`.

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -425,6 +568,21 @@ findPath(
 
+    double zScoreBuffer[MAX_PATHS];
+
+    for (int i = 0; i < bufferSize; i++) {
```
The Z-score is calculated after the fact and not taken into account during the path finding process right? Just to confirm that this is the case, otherwise we might want to push this up into the main loop.

Will-Tyler replied:
That is correct.

"Extension of the alignment path is based solely on the distance criteria. Neither the size of the gap, nor the statistical significance of the alignment path is considered at this point in the analysis."

JoaoRodrigues commented on Bio/PDB/ccealignmodule.c:
```diff
@@ -461,9 +619,11 @@ findPath(
             }
         }
 
+        const double zScore = zScoreBuffer[o];
```
Hm, with this getting a little more complex, I wonder if we shouldn't return a struct for each alignment with pathA, pathB, zscore, etc.

JoaoRodrigues commented on Bio/PDB/cealign.py:
```diff
@@ -110,36 +110,70 @@ def align(self, structure, transform=True):
         # corresponding RMSD. Use QCP to align the molecules.
-        best_rmsd, best_u = 1e6, None
-        for path in longest_paths:
+        aln = QCPSuperimposer()
```
Yeah, so, following my comment above. If we return a struct that can be converted to a namedtuple, we can simply keep the best namedtuple instead of carrying these variables around. Does this make sense?

I found [this](https://code.saghul.net/2012/04/pystructsequence-creating-named-tuples-in-c/) as a starting point to build a namedtuple in C, which might be nicer?

JoaoRodrigues commented on Bio/PDB/cealign.py:
```diff
@@ -110,36 +110,70 @@ def align(self, structure, transform=True):
+        if best_rigid_motion is None:
             raise RuntimeError("Failed to find a suitable alignment.")
 
+        # Gap optimization
```
I'd move this to a function of its own and add a kwarg that enables/disables this optimization from the top-level cealign call. 

Also, I'm surprised this is still so fast done in Python vs done in C!

JoaoRodrigues commented on Tests/test_PDB_CEAligner.py:
```diff
@@ -43,7 +43,7 @@ def test_cealigner(self):
         aligner.align(s2)
 
-        self.assertAlmostEqual(aligner.rms, 3.74, places=2)
+        self.assertAlmostEqual(aligner.rms, 3.66, places=2)
```
Yeah, so to my point above. I'd add this optimization as an optional, so that we can keep the 'raw' CEAlign code and tests vs the 'reference' implementation in Pymol, which is what most people use, and then add tests for the optimization itself.

JoaoRodrigues reviewed (changes requested):
Thank you super much @Will-Tyler, this is fantastic! A few comments mostly on code organization and documentation, but this looks great.

One thing that stood out: could you double check the lengths of the alignments are correct in your benchmarks? They're suspiciously the same as before!

Will-Tyler commented:
I think I addressed your review comments now. The alignment lengths are the same in the benchmark testing because the gap optimization does not change the length—I just included the length as a sanity check.

Since I made a lot of changes, I reran the test script and the results are still the same.

JoaoRodrigues commented on Bio/PDB/cealign.py:
```diff
@@ -86,7 +90,7 @@ def set_reference(self, structure):
             raise PDBException(msg)
 
-    def align(self, structure, transform=True):
+    def align(self, structure, transform=True, *, final_optimization=True):
```
Why the `*`? There were no additional arguments here so this is a safe add?

Will-Tyler replied:
The `*` makes the `final_optimization` parameter a keyword-only parameter (defined [here](https://docs.python.org/3/glossary.html#term-parameter)). I think it makes the code more clear if the caller has to do `aligner.align(structure, final_optimization=False)`.

And yes, because I am adding a new argument, the change does not break anyone's code.

JoaoRodrigues replied:
Oh gotcha, so that people won't do `align(struct, True True)`. Alright, makes sense!

JoaoRodrigues commented on Bio/PDB/cealign.py:
```diff
@@ -52,6 +52,10 @@ def __init__(self, window_size=8, max_gap=30):
         self.max_gap = max_gap
 
         self.rms = None
+        self._rigid_motion = None
```
I think this will be problematic if I align multiple structures to a reference. If one structure has a low RMSD, that structure's rotation/translation data will be used for all other ones. These variables should be local to the `align` function I think, just to be safe, or that function should reset them at every call.

Will-Tyler replied:
I don't think the scenario you described could happen because the code resets `self.rms` to `float("inf")` so the low RMSD from one alignment won't be reused, but I do think resetting the rigid motion data is a good idea. I will do that.

JoaoRodrigues replied:
I'd leave it as a local variable that is returned by optimize. It's a little annoying that the rmsd is also updated inside `_optimize`. I just worry about leftovers that cause problems with state with these variables.

Best case scenario was if the optimize function would return something like a named tuple or dataclass with the rotation/translation matrix and the rmsd, which you would then use to update the class variables but I understand if that's a big change. We also re-use the superimposer, which is again set as an instance variable, which has the same issue.

Up to you on what you think is best!

Will-Tyler commented:
I agree with you, and I prefer avoiding introducing state variables. I also don't like creating ad-hoc return types. In my first version of this pull request, I tried to avoid both of these by leaving the optimize code in the `align` function, but long functions are also a problem. I don't know of a quick fix for this, and I'm okay with it as it is.

JoaoRodrigues commented:
Thank you again @Will-Tyler ! Super nice contrib!

Will-Tyler commented:
Thanks for reviewing! 😄
